src/tsickle_compiler_host.ts

```diff
@@ -0,0 +1,199 @@
+  /** Converts a path to a module name */
+  pathToModuleName: (context: string, importPath: string) => string;
+  /** Tsickle treats warnings as errors, if true, ignore warnings */
+  shouldIgnoreWarningsForPath: (filePath: string) => boolean;
+}
+
+/**
+ * TsickleCompilerHost repeats the work of Google3CompilerHost, but with
+ * tsickle processing of input files, including closure type annotation
+ * processing, decorator downleveling and require -> googmodule rewriting.
+ */
+export class TsickleCompilerHost implements ts.CompilerHost {
+  private ANNOTATION_SUPPORT = `
+interface DecoratorInvocation {
+  type: Function;
+  args?: any[];
+}
+`;
+  // The manifest of JS modules output by the compiler.
+  public modulesManifest: ModulesManifest = new ModulesManifest();
+
+  /** Error messages produced by tsickle, if any. */
+  public diagnostics: ts.Diagnostic[] = [];
+
+  /** externs.js files produced by tsickle, if any. */
+  public externs: {[fileName: string]: string} = {};
+
+  constructor(
+      private delegate: ts.CompilerHost, private options: TsickleCompilerHostOptions,
+      private environment: TsickleEnvironment, private oldProgram?: ts.Program,
+      private pass = Pass.None) {}
+
+
+  public reconfigureForRun(program: ts.Program, pass: Pass) {
+    this.oldProgram = program;
+    this.pass = pass;
+  }
+
+  getSourceFile(
+      fileName: string, languageVersion: ts.ScriptTarget,
+      onError?: (message: string) => void): ts.SourceFile {
+    if (this.pass === Pass.None) {
+      return this.delegate.getSourceFile(fileName, languageVersion, onError);
+    }
+
+    if (!this.oldProgram) {
+      throw new Error('tried to run a pass other than None without setting a program');
+    }
+
+    const sourceFile = this.oldProgram.getSourceFile(fileName);
+    switch (this.pass) {
+      case Pass.DecoratorDownlevel:
+        return this.runDecoratorDownlevel(sourceFile, this.oldProgram, fileName, languageVersion);
+      case Pass.Tsickle:
+        return this.runTsickle(sourceFile, this.oldProgram, fileName, languageVersion);
+      default:
+        throw new Error('tried to use TsickleCompilerHost with unknown pass enum');
+    }
+  }
+
+  writeFile(
+      fileName: string, content: string, writeByteOrderMark: boolean,
+      onError?: (message: string) => void, sourceFiles?: ts.SourceFile[]): void {
+    // Strip any potential blaze-bin path (e.g. when compiling a generated
+    // file).
+    fileName = this.delegate.getCanonicalFileName(fileName);
+    if (this.options.googmodule && !fileName.match(/\.d\.ts$/)) {
+      content = this.convertCommonJsToGoogModule(fileName, content);
+    }
+
+    this.delegate.writeFile(fileName, content, writeByteOrderMark, onError, sourceFiles);
+  }
+
+  convertCommonJsToGoogModule(fileName: string, content: string): string {
+    // Strip off the file name extension.
+    fileName = fileName.substring(0, fileName.lastIndexOf('.'));
```

Comment:
As I recall one layer expects a filename with a .ts and another one doesn't.  It would be good to pin down which layer requires what (see my other comments about better documenting what the parameters are) to figure out whether this belongs here or inside g3.